File: Cpp/Modules/25.Printing and Exporting/20.CustomPreview/UPrinting.cpp
//---------------------------------------------------------------------------

#pragma hdrstop

#include "UPrinting.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

__fastcall TPrintingThread::TPrintingThread(TExcelFile *aXls, TFeedbackMethod aProgressFeedback,
	  TFinalFeedbackMethod aFinalFeedback, String aFileName, bool aAllSheets)
	  : TProgressThread(aXls, aProgressFeedback, aFinalFeedback, aFileName, aAllSheets)
{
}


void __fastcall TPrintingThread::Execute()
{
  TFlexCelPrintDocument* doc = new TFlexCelPrintDocument(Xls);
  try
  {
	doc->AfterGeneratePage = ShowProgress;

	if (AllSheets)
	{
		doc->BeginPrint();
		try
		{
		  doc->PrintAllVisibleSheets(false);
		}
		__finally
		{
		  doc->EndPrint();
		}
	}
	else
	{
	  doc->Print();
	}
  }
  __finally
  {
	doc->Free();
  }
}

#if (__BORLANDC__ < 0x0650) //XE2 or older
//C++ builder header translations are wrong for XE2 or older
//see http://qc.embarcadero.com/wc/qcmain.aspx?d=42782
void __fastcall TPrintingThread::ShowProgress(const TObject* sender, const TPrintPageEventArgs* e)
#else
void __fastcall TPrintingThread::ShowProgress(TObject* const sender, TPrintPageEventArgs* const e)
#endif
{
  TFlexCelPrintingProgress* Prog = ((TFlexCelPrintDocument*)sender)->Progress;
  int Percent;
  if (Prog->TotalPage == 0) Percent = 100; else Percent = (int)(Prog->Page * 100.0 / Prog->TotalPage);

  ProgressFeedback_percent = Percent;
  ProgressFeedback_msg = String(L"Page ") + IntToStr(Prog->Page) + L" of " + IntToStr(Prog->TotalPage);
  Synchronize(&DoProgressFeedback);
}

File: Delphi/Modules/25.Printing and Exporting/20.CustomPreview/UPrinting.pas
unit UPrinting;

interface
uses UProgressThread, SysUtils, FlexCel.Core, FlexCel.Render, Classes;

type
  TPrintingThread = class(TProgressThread)
  private
    FirstPage, TotalPages: integer;
    procedure ShowProgress(const sender: TObject; const e: TPrintPageEventArgs);
  protected
    procedure Execute; override;
  public
    constructor Create(const aXls: TExcelFile; const aProgressFeedback: TProc<integer, string>;
    const aFinalFeedback: TProc<boolean, string>; const aFileName: string;
    const aAllSheets: boolean; const aFirstPage, aTotalPages: integer);
  end;


implementation

{ TPrintingThread }

constructor TPrintingThread.Create(const aXls: TExcelFile;
  const aProgressFeedback: TProc<integer, string>;
  const aFinalFeedback: TProc<boolean, string>; const aFileName: string;
  const aAllSheets: boolean; const aFirstPage, aTotalPages: integer);
begin
  inherited Create(aXls, aProgressFeedback, aFinalFeedback, aFileName, aAllSheets);
  FirstPage := aFirstPage;
  TotalPages := aTotalPages;
end;

procedure TPrintingThread.Execute;
var
  doc: TFlexCelPrintDocument;
begin
  inherited;
  doc := TFlexCelPrintDocument.Create(Xls);
  try
    doc.AfterGeneratePage := ShowProgress;
    if AllSheets then
    begin
      doc.BeginPrint;
      try
        doc.PrintAllVisibleSheets(false, FirstPage, TotalPages);
      finally
        doc.EndPrint;
      end;
    end else
    begin
      doc.BeginPrint;
      try
        doc.PrintSheet(1, -1, FirstPage, TotalPages);
      finally
        doc.EndPrint;
      end;
    end;
  finally
    FreeAndNil(doc);
  end;
end;

procedure TPrintingThread.ShowProgress(const sender: TObject;
  const e: TPrintPageEventArgs);
var
  Prog: TFlexCelPrintingProgress;
  Percent: Integer;
begin
  Prog := (Sender as TFlexCelPrintDocument).Progress;
  if (Prog.TotalPage = 0) then Percent := 100 else Percent := Round(Prog.Page * 100.0 / Prog.TotalPage);

  Synchronize(
  procedure
  begin
    ProgressFeedback(Percent, 'Page ' + IntToStr(Prog.Page) + ' of ' + IntToStr(Prog.TotalPage));
  end);
end;

end.
